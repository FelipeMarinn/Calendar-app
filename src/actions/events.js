import Swal from "sweetalert2";
import { fetchConToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";


export const eventStartAddNew = ( event ) => {
    return async( dispatch, getState ) => {

        const { uid, name } = getState().auth

        try {

            const resp = await fetchConToken('events', event, 'POST')
            const body = await resp.json()
    
            if ( body.ok ) {

                event.id = body.event.id
                event.user = {
                    _id: uid,
                    name: name
                }
                console.log(event)
                dispatch( eventAddNew( event ) )
            }

        } catch (err) {
            console.log(err)
        }
    
    }
}

export const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
})

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
})

export const eventClearActiveEvent = () => ({ type: types.eventClearActiveEvent })


export const eventStartUpdate = ( event ) => {
    return async( dispatch ) => {

        try {

            const resp = await fetchConToken(`events/${ event.id }`, event, 'PUT')
            const body = await resp.json()

            if ( body.ok ) {
                dispatch( eventUpdated(event) )
            } else {
                Swal.fire('Error', body.msg, 'error')
            }

        } catch (err) {
            console.log(err)
        }
    }
}
const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
})


export const eventStartDelete = () => {
    return async( dispatch, getState ) => {

        const { id } = getState().calendar.activeEvent
        console.log( id )
        try {

            const resp = await fetchConToken(`events/${ id }`, {}, 'DELETE')
            const body = await resp.json()

            if ( body.ok ) {
                dispatch( eventDeleted() )
            } else {
                Swal.fire('Error', body.msg, 'error')
            }

        } catch (err) {
            console.log(err)
        }

    }
}
const eventDeleted = () => ({ type: types.eventDelected })


export const eventStartLoading = () => {
    return async( dispatch ) => {

        try {

            const resp = await fetchConToken( 'events' )
            const body = await resp.json()

            const events = prepareEvents( body.events )
            dispatch( eventLoaded(events) )

        } catch (err) {
            console.log(err)
        }
    }
}

const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
})

export const eventLogout = () => ({ type: types.eventLogout })