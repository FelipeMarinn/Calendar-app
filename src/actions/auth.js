import { fetchConToken, fetchSinToken } from "../helpers/fetch"
import { types } from "../types/types"
import Swal from 'sweetalert2'
import { eventLogout } from "./events"


export const startLogin = ( email, password ) => {
    return async( dispatch ) => {

        const resp = await fetchSinToken( 'auth', { email, password }, 'POST' )
        const body = await resp.json()
        
        if ( body.ok ) {
            localStorage.setItem('token', body.token)
            localStorage.setItem('token-init-date', new Date().getTime())
            
            dispatch( login({
                uid: body.uid,
                name: body.name
            }) )
        } else {
            Swal.fire('Error', body.msg, 'error')
        }
    }
}

export const startRegister = (  name, email, password ) => {
    return async( dispatch ) => {
        
        const resp = await fetchSinToken( 'auth/new', { name, email, password }, 'POST' )
        const body = await resp.json()
        
        if ( body.ok ) {
            localStorage.setItem('token', body.token)
            localStorage.setItem('token-init-date', new Date().getTime())
            
            dispatch( login({
                uid: body.uid,
                name: body.name
            }) )
        } else {
            Swal.fire('Error', body.msg, 'error')
        }

    }
}

export const startChecking = () => {
    return async( dispatch ) => {

        try {
            const resp = await fetchConToken( 'auth/renew' )
            const body = await resp.json()
    
            if ( body.ok ) {
                localStorage.setItem('token', body.token)
                localStorage.setItem('token-init-date', new Date().getTime())
                
                dispatch( login({
                    uid: body.uid,
                    name: body.name
                }) )
            } else {
                dispatch( checkingFinish() )
            }
        }  catch (err) {
            console.log('Error:', err )
            dispatch( sistemError() )
        }

    }
}

const sistemError = () => ({ type: types.sistemError })
 
const checkingFinish = () => ({ type: types.authCheckingFinish })

const login = ( user ) => ({
    type: types.authLogin,
    payload: user
})

export const startLogout = () => {
    return ( dispatch ) => {

        localStorage.clear()
        dispatch( logout() )
        dispatch( eventLogout() )

    }
}

const logout = () => ({ type: types.authLogout })


