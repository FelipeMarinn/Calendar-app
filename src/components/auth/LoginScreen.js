import React from 'react';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogin } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import './login.css';


export const LoginScreen = () => {

    const dispatch = useDispatch()

    const [ formValues, handleInputChange ] = useForm({
        email: 'daniel@gmail.com',
        password: '123456'
    });

    const { email, password } = formValues

    const handleLogin = (e) => {
        e.preventDefault()      
        dispatch( startLogin( email, password ) )
    }


    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 m-auto text-center login-form-1">
                    <h3>Ingreso</h3>
                    <form onSubmit={ handleLogin }>
                        <div className="form-group">
                            <input 
                                type="text"
                                className="form-control"
                                placeholder="Correo"
                                name="email"
                                value={ email }
                                onChange={ handleInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="ContraseÃ±a"
                                name="password"
                                value={ password }
                                onChange={ handleInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input 
                                type="submit"
                                className="btnSubmit"
                                value="Login" 
                            />
                        </div>
                    </form>
                    <Link to='/register' > or register </Link>
                </div>
            </div>
        </div>
    )
}