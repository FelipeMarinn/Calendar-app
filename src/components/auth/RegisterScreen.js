import React from 'react'
import { useDispatch } from 'react-redux'
import { Link } from 'react-router-dom'
import Swal from 'sweetalert2'
import { startRegister } from '../../actions/auth'
import { useForm } from '../../hooks/useForm'


export const RegisterScreen = () => {
    
    const dispatch = useDispatch()

    const [ formValues, handleInputChange ] = useForm({
        name: 'Andres',
        email: 'andres@gmail.com',
        password1: '123456',
        password2: '123456'
    });

    const { name, email, password1, password2 } = formValues

    const handleRegister = (e) => {
        e.preventDefault()    
        
        if ( password1 !== password2 ) {
            return Swal.fire('Error', 'Las contraseñas no coinciden', 'error')
        }
        
        dispatch( startRegister( name, email, password1 ) )
    }


    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 text-center m-auto login-form-2">
                    <h3>Registro</h3>
                    <form onSubmit={ handleRegister }>
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Nombre"
                                name='name'
                                value={ name }
                                onChange={ handleInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Correo"
                                name='email'
                                value={ email }
                                onChange={ handleInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contraseña" 
                                name='password1'
                                value={ password1 }
                                onChange={ handleInputChange }
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Repita la contraseña" 
                                name='password2'
                                value={ password2 }
                                onChange={ handleInputChange }
                            />
                        </div>

                        <div className="form-group">
                            <input 
                                type="submit" 
                                className="btnSubmit" 
                                value="Crear cuenta" />
                        </div>
                    </form>
                    <Link to='/login' className='text-white'> Sign in </Link>
                </div>
            </div>
        </div>
    )
}
