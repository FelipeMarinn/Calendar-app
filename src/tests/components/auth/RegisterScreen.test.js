import React from 'react'
import { mount } from 'enzyme'
import { Provider } from 'react-redux'
import Swal from 'sweetalert2'

import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { MemoryRouter } from 'react-router'
import { RegisterScreen } from '../../../components/auth/RegisterScreen'
import { startRegister } from '../../../actions/auth'


jest.mock('../../../actions/auth', () => ({
    startRegister: jest.fn()
}))
jest.mock('sweetalert2', () => ({
    fire: jest.fn()
}))

const middlewares = [thunk]
const mockStore = configureStore( middlewares )

const initState = {}
const store = mockStore( initState )
store.dispatch = jest.fn()

const wrapper = mount(
    <Provider store={ store }>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
)


describe('RegisterScreen', ()=> {

    beforeEach(() => {
        jest.clearAllMocks()
    })

    test('se debe de renderizar correctamente', () => {
        expect( wrapper ).toMatchSnapshot()
    })

    test('No hay registro si las contraseñas son diferentes', ()=> {
        
        wrapper.find('input[name="password1"]').simulate('change', {
            target: {
                name: 'password1',
                value: '000125'
            }
        })
        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        })

        expect( startRegister ).not.toHaveBeenCalled()
        expect( Swal.fire ).toHaveBeenCalledWith('Error', 'Las contraseñas no coinciden', 'error')
        
    })

    test('Registro con contraseñas iguales', () => {

        wrapper.find('input[name="password1"]').simulate('change', {
            target: {
                name: 'password1',
                value: '123456'
            }
        })
        wrapper.find('input[name="password2"]').simulate('change', {
            target: {
                name: 'password2',
                value: '123456'
            }
        })

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        })

        expect( Swal.fire ).not.toHaveBeenCalled()
        expect( startRegister ).toHaveBeenCalledWith('Andres', 'andres@gmail.com', '123456')
    })

})