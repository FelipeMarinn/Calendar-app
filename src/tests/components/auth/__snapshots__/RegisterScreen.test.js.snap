// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterScreen se debe de renderizar correctamente 1`] = `
<div
  className="container login-container register"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 text-center m-auto login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="name"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="Andres"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="email"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="andres@gmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="password1"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="password2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
      <a
        className="text-white"
        href="/login"
        onClick={[Function]}
      >
         Sign in 
      </a>
    </div>
  </div>
</div>
`;
